### PR TITLE
Implement GSensor support for more moderm PocketBooks

### PR DESCRIPTION
**Pull request by rjd22**
Requires: https://github.com/koreader/koreader-base/pull/1562 and https://github.com/koreader/koreader-base/pull/1566
Issue: #9427

~~I feel I will need to ask for a lot of forgiveness now after writing this stuff. Most likely this implementation is completely in the wrong place and implemented completely in the wrong way but it actually works.~~

~~Because the lack of a proper event when the screen is rotated (I've looked, and looked, and looked even more but didn't find one). The only way I know ATM is polling the GSensor every second and just broadcasting an event for the screen change when there was a change found.~~

~~Most likely this is extremely inefficient so feel free to point me to a better way of doing this. But I was not able to find some more guiding implementation on the gyro when it is not event based.~~

~~Only enabled on my device for now because I don't know what devices have the more modern API and am unable to test this code on all of them.~~

~~I'm sorry for the mess :>~~

Now with the event based implementation. This is tested on my PocketBook and working nicely.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9873)

### CODE REVIEW

**Comment by NiLuJe** [comment]
Yeah, this should definitely be spun up in a dedicated thread/process in base, one that injects custom input events, like we do for lipc on Kindle, or uevent elsewhere.

e.g., see https://github.com/koreader/koreader-base/blob/master/input/input-kobo.h

**Comment by rjd22** [comment]
@NiLuJe sounds like a good idea. Where would this go for Pocketbook? 

The whole event implementation seems to be in FFI and hacking it into inkview-compat doesn't sound like a great idea ^^°.

**Comment by NiLuJe** [comment]
Since PB already has an input module, you'd need to *also* load this C module and run it from Device:init.

You can probably just stuff new functions inside the existing C pocketbook module, that'll save you dealing with the buildsystem ;).

We usually handle this via `self.input.open("fake_events")`, but you can just export the function and call that directly, since we no longer really use the rest of the C PB input module.

**Comment by NiLuJe** [comment]
On the other hand, the PB InkView mess means you'll possibly have to jump some extra hoops as you don't really have an evdev device to inject events into right now...

**Comment by NiLuJe** [comment]
Sooo, yeah, probably just easier to shove it into ffi/input_pocketbook.lua or inkview-compat... somewhere :/.

**Comment by NiLuJe** [comment]
Possibly in `translateeEvent`...

On the other hand, I'm mildly confused as to why you're not simply seeing an `EVT_ORIENTATION`? Possibly missing the init_gsensor stuff?

**Comment by rjd22** [comment]
@NiLuJe I'm also really unsure why I don't see those events. I got still no response from pocketbook and there is little documentation on the gsensor.

I added a debug statement on top of translateEvent in the ffi implementation but didn't see any events passing along when turning the device with the gyro enabled (the power led turns on when turning the device so it responds to the turning by pulling the device out of sleep).

Would there be drawbacks in placing the gyro code in the ffi implementation? I could use inkview-compat to fix the symbol lookup failing (i feel this is the way).

**Comment by NiLuJe** [comment]
I'd rather figure out why EVT_ORIENTATION doesn't fire, as this seems to be The Way, looking at a quick GH search ;).

See if forcing the compat PrepareForLoop symbol helps or not, as well as forcing the iv_setup stuff, maybe?

**Comment by NiLuJe** [comment]
(Also, pcall works for the symbol lookup, forgot about that, see the compat stuff ;p).

**Comment by rjd22** [comment]
> I'd rather figure out why EVT_ORIENTATION doesn't fire, as this seems to be The Way, looking at a quick GH search ;).

Might it not be that is triggered by the `setOrientation` inkview call and actually has nothing to do with the gyro  (they are related but separate it seems).

 https://github.com/pocketbook/SDK_6.3.0/blob/6.5/SDK-B288/usr/arm-obreey-linux-gnueabi/sysroot/usr/local/include/inkview.h#L1726

**Comment by NiLuJe** [comment]
That wouldn't make much sense ;). I'd imagine setOrientation is to programmatically force an UI orientation change, not the other way around.

**Comment by rjd22** [comment]
> > See if forcing the compat PrepareForLoop symbol helps or not, as well as forcing the iv_setup stuff, maybe?
> 
> In the same vein, I'd check if you actually see the events when using the C module, on the off-chance `PrepareForLoop` does things slightly differently?

Tried to set the boolean true but it resulted in a crash. It seems the C module is not working anymore ^^". I'm going to try the other things too.

**Comment by rjd22** [comment]
> I'd rather figure out why EVT_ORIENTATION doesn't fire, as this seems to be The Way, looking at a quick GH search ;).
> 
> See if forcing the compat PrepareForLoop symbol helps or not, as well as forcing the iv_setup stuff, maybe?

I've just tried this and no extra events. I've also tried various combinations of the following calls to see if it triggers something to return the event:

```lua
    inkview.hw_init()
    inkview.iv_update_orientation(-1)
    inkview.iv_setup_touchpanel()
    inkview.iv_setup_gsensor()
```

I've tried setting the `SetGSensor` call to both `1` and `2` and `SetGlobalOrientation` to `-1` but still no orientation events.

This is the debug code I'm using in `input_pocketbook.lua` to debug the event. Sadly I'm unable to debug with the `.c` side of things because it seems broken. Note the debug statements added:

```lua
-- Translate event from inkview EVT_* into emulated linuxev EV_
local function translateEvent(t, par1, par2)
    logger.dbg("rj event: ", t, par1, par2)

    if eventq == nil then
        return 0
    end
    updateTimestamp()
    if t == C.EVT_INIT then
        inkview.SetPanelType(C.PANEL_DISABLED);
    elseif t == C.EVT_ORIENTATION then
        logger.dbg("orientation event: ", t, par1, par2)
    elseif t == C.EVT_POINTERDOWN then
```

Attached is my debug log showing that I'm receiving other input events but no orientation events. Even when the gyro poll shows that the gyro is on and is registering different poll values.

EVT_ORIENTATION should be of code `32`

[crash.log](https://github.com/koreader/koreader/files/10165146/crash.log)

**Comment by rjd22** [comment]
I also saw this comment in the pocketbook source https://github.com/pocketbook/SDK_6.3.0/blob/6.5/SDK-B288/usr/arm-obreey-linux-gnueabi/sysroot/usr/local/include/inkinternal.h#L469 . It might be that with newer inkview versions the `EVT_ORIENTATION` might be getting filtered now. 

Sadly trying to change the mode doesn't do anything. I tried changing the mode with both the `SetGSensor` and `hw_set_gsensor` but still no joy.

This kind of concludes all the research I can do with my technical capabilities. So if someone with a Pocketbook device wants to help I would gladly take it right now.

**Comment by rjd22** [comment]
> I'm mildly interested in the crash info, because ideally, the C module shouldn't do that ;p.

I will make a crash report for you so you can check it when I have some more time. But it has to do with the `input()` call in pocketbook/device.lua not being compatible with the c input call for Pocketbook at least. Too much has changed after FFI was implemented.

> But, back to the issue at hand:
> 
>     * I'd actually implement the EVT_ORIENTATION code, as it can't really hurt
> 
>     * I'd fudge the polling somewhere in there, possibly in translateEvent by checking the timestamp delta and doing it every couple seconds? Possibly only after a SYN_REPORT to avoid confusing the input code?

I was wondering if I could use fakeEvent for this since it seems meant for situations like this.

The implementation I thought up in my head was to introduce a polling mechanism outside of translateEvent that gets booted by input:open. It keeps polling and passing the events directly to genEmuEvent until the first EVT_ORIENTATION event passes along thought translateEvent.

This way if the event mechanism works correctly or starts working correctly the polling mechanism gets cancelled. If it doesn't work correctly the polling mechanism stays alive. I could make use of inkview.weakTimer for the polling mechanism unless you have a better method of polling with basic lua.

Sadly backwards compatibility with older devices won't work for at least the polling mechanism because the old gyro functions also don't seem to work as expected ^^°.

In any case. It's a big mess.

**Comment by NiLuJe** [comment]
Which fakeEvent are we talking about? (not on my dev box, so I don't have the code easily searchable).

If you can defer polling to InkView somehow, sure, go for it. I was sorta thinking of bypassing the polling approach altogether and instead piggybacking on actual events (and possibly just input events), which is why I mentioned just checking the last timestamp after every SYN_REPORT to decide whether to poke the gryo or not ;).

**Comment by rjd22** [comment]
> Which fakeEvent are we talking about? (not on my dev box, so I don't have the code easily searchable).

I meant this part: https://github.com/koreader/koreader-base/blob/master/ffi/input_pocketbook.lua#L293

In the c implementation of kobo there was something similar but had less specific naming for generating not real events.

> If you can defer polling to InkView somehow, sure, go for it. I was sorta thinking of bypassing the polling approach altogether and instead piggybacking on actual events (and possibly just input events), which is why I mentioned just checking the last timestamp after every SYN_REPORT to decide whether to poke the gryo or not ;).

Sadly piggybacking won't be possible AFAIK. Pocketbook's are extremely silent when you don't touch them. Even sleeping and waking up doesn't throw any events. For my device the only events that seem to be thrown are button and touch events.

**Comment by NiLuJe** [comment]
> Sadly piggybacking won't be possible AFAIK. Pocketbook's are extremely silent when you don't touch them. Even sleeping and waking up doesn't throw any events. For my device the only events that seem to be thrown are button and touch events.

I know, this would essentially delay poking the gyro until the next user input, which felt like a somewhat fair trade if it allowed not making a mess of the code ;p.

**Comment by rjd22** [comment]
> > Sadly piggybacking won't be possible AFAIK. Pocketbook's are extremely silent when you don't touch them. Even sleeping and waking up doesn't throw any events. For my device the only events that seem to be thrown are button and touch events.
> 
> I know, this would essentially delay poking the gyro until the next user input, which felt like a somewhat fair trade if it allowed not making a mess of the code ;p.

In my usecase I often get an image that is too wide for portrait so I have to turn the device. Having to move a page and back for the gyro to work would be no different from the turn screen gesture :p.

**Comment by NiLuJe** [comment]
That's a fair point ;o) (I have mine disabled ^^).

**Comment by rjd22** [comment]
@NiLuJe I've implemented the gsensor with the new firmware and information I got from PocketBook and it's working nicely. What do you think of the implementation?

**Comment by NiLuJe** [comment]
I would much rather have the PB FFI stuff in base properly translate that to an EV_MSC (with a custom code if need be), instead of keeping this as a custom *type* and mangling the common codepaths in front to deal with it ;).

**Comment by NiLuJe** [comment]
Note to self: In fact, do the same for the Kindle stuff, because that ABS hack is nasty.

**Comment by rjd22** [comment]
@NiLuJe that would be indeed be a lot better. Could you point me to the part of the input.lua that is the official rotation code so I know what the events look like? They all looked like custom implementations to me.

**Comment by NiLuJe** [comment]
They're unfortunately all custom-ish, but, yeah, follow the Kobo one ;).

**Comment by rjd22** [comment]
> They're unfortunately all custom-ish, but, yeah, follow the Kobo one ;).

Imo it would be better in the future it split them all up into separate input modules for every platform and having them talk if an more abstract api. Translating events into other events that then are converted to method calls are a little hard to follow.

**Comment by rjd22** [comment]
@NiLuJe I checked if I could use the toggleMiscEvNTX handler but it seems with PocketBook this will be quite a hassle because of the deep dependency with Kobo and incompatibilities of EV_MSC and PocketBook. It looks like it would require me to rip out the whole rooted PocketBook code that uses EV_MSC for synthetic Power events.

https://github.com/koreader/koreader/blob/master/frontend/device/input.lua#L1404-L1413

Also the toggling code used by kobo is not compatible with how Pocketbook handles things:

https://github.com/koreader/koreader/blob/master/frontend/device/input.lua#L995-L1018
https://github.com/koreader/koreader/blob/master/frontend/device/input.lua#L995-L1018

This is most likely also the reason with there is a different kindle handler. How do you want me to proceed?

**Comment by NiLuJe** [comment]
Yeah, forgot that PB already had an EV_MSC handler... I'll think on it a bit ;o).

**Comment by NiLuJe** [comment]
Okay, went with a refactor and a move to our own protocol to simplify the Input/Generic codepaths.

c.f., https://github.com/NiLuJe/koreader-base/tree/gyro-refactor + https://github.com/NiLuJe/koreader/tree/gyro-refactor ;).

I've wired the protocol handling in PocketBook, it just needs the `hasGSensor` Device cap (and possibly your custom toggleGSensor implementation, if necessary?).

**Comment by rjd22** [comment]
@NiLuJe I've checked your changes ant that looks great! Thanks a lot!

I will check what is still needed for PocketBook hopefully in the coming days.

**Review by NiLuJe — APPROVED** [review]

**Comment by rjd22** [comment]
@NiLuJe thx! I bumped base it's GTG

**Comment by rjd22** [comment]
That was quite a ride! Thanks a lot for all the help :D

**Comment by NiLuJe** [comment]
Thanks for dealing so good-naturedly with my initial ramblings when I didn't even quite remember how everything involved worked ^^.
